radixdlt/src/main/java/com/radixdlt/middleware2/network/TestEventCoordinatorNetwork.java

```diff
@@ -115,87 +104,80 @@ public void setReceivingDisable(ECPublicKey validatorId, boolean disable) {
 		}
 	}
 
-	private int getRandomLatency() {
-		if (minimumLatency == maximumLatency) {
-			return minimumLatency;
-		} else {
-			return minimumLatency + rng.nextInt(maximumLatency - minimumLatency + 1);
-		}
-	}
-
 	public EventCoordinatorNetworkSender getNetworkSender(ECPublicKey forNode) {
-		Consumer<MessageInTransit> sendMessageSink = message -> {
-			if (!sendingDisabled.contains(forNode)) {
-				if (preserveOrder) {
-					orderedMessageBuffer.push(message);
-				}
-				executorService.schedule(() -> {
-					if (preserveOrder) {
-						MessageInTransit otherMessage = orderedMessageBuffer.pollLast();
-						if (otherMessage == null) {
-							throw new IllegalStateException("No message available in message buffer");
-						}
-						receivedMessages.onNext(otherMessage);
-					} else {
-						receivedMessages.onNext(message);
-					}
-				}, getRandomLatency(), TimeUnit.MILLISECONDS);
-			}
-		};
 		return new EventCoordinatorNetworkSender() {
 			@Override
 			public void broadcastProposal(Proposal proposal) {
-				sendMessageSink.accept(MessageInTransit.broadcast(proposal));
+				for (ECPublicKey reader : readers) {
+					receivedMessages.onNext(MessageInTransit.send(proposal, forNode, reader));
+				}
 			}
 
 			@Override
 			public void sendNewView(NewView newView, ECPublicKey newViewLeader) {
-				sendMessageSink.accept(MessageInTransit.send(newView, newViewLeader));
+				receivedMessages.onNext(MessageInTransit.send(newView, forNode, newViewLeader));
 			}
 
 			@Override
 			public void sendVote(Vote vote, ECPublicKey leader) {
-				sendMessageSink.accept(MessageInTransit.send(vote, leader));
+				receivedMessages.onNext(MessageInTransit.send(vote, forNode, leader));
 			}
 		};
 	}
 
 	public EventCoordinatorNetworkRx getNetworkRx(ECPublicKey forNode) {
+		readers.add(forNode);
 		// filter only relevant messages (appropriate target and if receiving is allowed)
-		Observable<Object> myMessages = receivedMessages
-			.filter(message -> !receivingDisabled.contains(forNode))
-			.filter(message -> message.isRelevantFor(forNode))
-			.map(MessageInTransit::getContent);
-		return () -> myMessages.ofType(ConsensusEvent.class);
+		Observable<ConsensusEvent> myMessages = receivedMessages
+			.filter(msg -> !sendingDisabled.contains(msg.sender))
+			.filter(msg -> !receivingDisabled.contains(msg.target))
+			.filter(msg -> msg.target.equals(forNode))
+			.timestamp(TimeUnit.MILLISECONDS)
+			.scan((msg1, msg2) -> {
+				if (msg2.value().sender.equals(forNode)) {
+					return msg2;
+				}
+				int delayCarryover = (int) Math.max(msg1.time() + msg1.value().delay - msg2.time(), 0);
+				int range = maximumLatency - delayCarryover - minimumLatency + 1;
+				// Currently a gaussian distribution
+				int nextDelay = range > 0 ? minimumLatency + rng.nextInt(range) : 0;
```

Comment:
Should this be `range > 0 ? minimumLatency + rng.nextInt(range) : Math.max(0, minimumLatency + range)`?

Reply:
Or maybe `range > 0 ? minimumLatency + rng.nextInt(range) : minimumLatency`?

Reply:
I think your first comment is right

Reply:
or rather it should just be: `range > 0 ? minimumLatency + rng.nextInt(range) : minimumLatency + range`